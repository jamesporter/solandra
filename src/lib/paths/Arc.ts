import { Point2D } from "../types/sol"
import { Traceable } from "./index"
export class Arc implements Traceable {
  readonly cX: number
  readonly cY: number
  readonly radius: number
  readonly startAngle: number
  readonly endAngle: number
  readonly antiClockwise: boolean
  constructor(config: { at: Point2D; r: number; a: number; a2: number }) {
    const {
      at: [cX, cY],
      r,
      a,
      a2,
    } = config
    const antiClockwise = a > a2
    this.cX = cX
    this.cY = cY
    this.radius = r
    this.startAngle = a
    this.endAngle = a2
    this.antiClockwise = antiClockwise
  }
  traceIn = (ctx: CanvasRenderingContext2D) => {
    if (Math.abs(this.startAngle - this.endAngle) > 0.0001)
      ctx.moveTo(this.cX, this.cY)
    ctx.arc(
      this.cX,
      this.cY,
      this.radius,
      this.startAngle,
      this.endAngle,
      this.antiClockwise
    )
    if (this.startAngle - this.endAngle > 0.0001) ctx.lineTo(this.cX, this.cX)
  }
}
