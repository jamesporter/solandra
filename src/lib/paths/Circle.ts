import { Point2D } from "../types/sol"
import { Ellipse } from "."

/**
 * Just an ellipse with width = height
 */
export class Circle extends Ellipse {
  constructor(config: {
    at: Point2D
    r: number
    align?: "center" | "topLeft"
  }) {
    super({
      at: config.at,
      w: config.r * 2,
      h: config.r * 2,
      align: config.align,
    })
  }
}
