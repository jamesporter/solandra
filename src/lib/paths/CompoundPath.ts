import { Traceable } from "."

export class CompoundPath implements Traceable {
  private constructor(private paths: Traceable[]) {}

  static withPaths(...paths: Traceable[]) {
    return new CompoundPath(paths)
  }

  traceIn(ctx: CanvasRenderingContext2D) {
    this.paths.forEach((p) => p.traceIn(ctx))
  }
}
