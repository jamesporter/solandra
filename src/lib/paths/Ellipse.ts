import { Point2D } from "../types/sol"
import { Traceable } from "./index"
/**
 * Technically you can't do ellipses/circles properly with cubic beziers, but you can come very, very close
 *
 * Uses 4 point, cubic beziers, approximation of (4/3)*tan(pi/8) for control points
 *
 * https://stackoverflow.com/questions/1734745/how-to-create-circle-with-bÃ©zier-curves
 */
export class Ellipse implements Traceable {
  constructor(
    protected config: {
      at: Point2D
      w: number
      h: number
      align?: "center" | "topLeft"
    }
  ) {}
  traceIn = (ctx: CanvasRenderingContext2D) => {
    const { at, w: width, h: height, align = "center" } = this.config
    const [cX, cY] =
      align === "center" ? at : [at[0] + width / 2, at[1] + height / 2]
    const a = (4 / 3) * Math.tan(Math.PI / 8)
    ctx.moveTo(cX, cY - height / 2)
    ctx.bezierCurveTo(
      cX + (a * width) / 2,
      cY - height / 2,
      cX + width / 2,
      cY - (a * height) / 2,
      cX + width / 2,
      cY
    )
    ctx.bezierCurveTo(
      cX + width / 2,
      cY + (a * height) / 2,
      cX + (a * width) / 2,
      cY + height / 2,
      cX,
      cY + height / 2
    )
    ctx.bezierCurveTo(
      cX - (a * width) / 2,
      cY + height / 2,
      cX - width / 2,
      cY + (a * height) / 2,
      cX - width / 2,
      cY
    )
    ctx.bezierCurveTo(
      cX - width / 2,
      cY - (a * height) / 2,
      cX - (a * width) / 2,
      cY - height / 2,
      cX,
      cY - height / 2
    )
  }
}
