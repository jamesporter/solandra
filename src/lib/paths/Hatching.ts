import { Point2D } from "../types/sol"
import { Traceable } from "./index"
/**
 * Hatching in a circle around a point, with a radius and delta between lines
 */
export class Hatching implements Traceable {
  constructor(
    private config: {
      at: Point2D
      r: number
      a: number
      delta: number
    }
  ) {}
  traceIn = (ctx: CanvasRenderingContext2D) => {
    const {
      at: [x, y],
      r,
      a,
      delta,
    } = this.config
    let perpOffset = 0
    const rca = r * Math.cos(a - Math.PI / 2)
    const rsa = r * Math.sin(a - Math.PI / 2)
    const dX = Math.cos(a)
    const dY = Math.sin(a)
    ctx.moveTo(x - rca, y - rsa)
    ctx.lineTo(x + rca, y + rsa)
    while (perpOffset < r) {
      perpOffset += delta
      const [sX, sY] = [perpOffset * dX, perpOffset * dY]
      // divide by r as using rsa/rca below
      const rl = Math.sqrt(r * r - perpOffset * perpOffset) / r
      ctx.moveTo(x + sX - rl * rca, y + sY - rl * rsa)
      ctx.lineTo(x + sX + rl * rca, y + sY + rl * rsa)
      ctx.moveTo(x - sX - rl * rca, y - sY - rl * rsa)
      ctx.lineTo(x - sX + rl * rca, y - sY + rl * rsa)
    }
  }
}
