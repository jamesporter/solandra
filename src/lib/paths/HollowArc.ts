import { Point2D } from "../types/sol"
import { Traceable } from "./index"
export class HollowArc implements Traceable {
  readonly cX: number
  readonly cY: number
  readonly radius: number
  readonly innerRadius: number
  readonly startAngle: number
  readonly endAngle: number
  readonly antiClockwise: boolean
  constructor(config: {
    at: Point2D
    r: number
    r2: number
    a: number
    a2: number
  }) {
    const {
      at: [cX, cY],
      r,
      r2,
      a,
      a2,
    } = config
    const antiClockwise = a > a2
    this.cX = cX
    this.cY = cY
    this.radius = r
    this.innerRadius = r2
    this.startAngle = a
    this.endAngle = a2
    this.antiClockwise = antiClockwise
  }
  traceIn = (ctx: CanvasRenderingContext2D) => {
    ctx.moveTo(
      this.cX + this.innerRadius * Math.cos(this.startAngle),
      this.cY + this.innerRadius * Math.sin(this.startAngle)
    )
    ctx.lineTo(
      this.cX + this.radius * Math.cos(this.startAngle),
      this.cY + this.radius * Math.sin(this.startAngle)
    )
    ctx.arc(
      this.cX,
      this.cY,
      this.radius,
      this.startAngle,
      this.endAngle,
      this.antiClockwise
    )
    ctx.lineTo(
      this.cX + this.innerRadius * Math.cos(this.endAngle),
      this.cY + this.innerRadius * Math.sin(this.endAngle)
    )
    ctx.arc(
      this.cX,
      this.cY,
      this.innerRadius,
      this.endAngle,
      this.startAngle,
      !this.antiClockwise
    )
  }
}
