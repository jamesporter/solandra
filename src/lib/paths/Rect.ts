import { Point2D } from "../types/sol"
import { Traceable } from "."
import { v } from ".."
import { SimplePath } from "./SimplePath"
export class Rect implements Traceable {
  readonly at: Point2D
  readonly w: number
  readonly h: number

  constructor(config: {
    at: Point2D
    w: number
    h: number
    align?: "topLeft" | "center"
  }) {
    const { at, w, h, align = "topLeft" } = config
    this.at = align === "topLeft" ? at : v.subtract(at, [w / 2, h / 2])
    this.w = w
    this.h = h
  }

  traceIn = (ctx: CanvasRenderingContext2D) => {
    ctx.rect(this.at[0], this.at[1], this.w, this.h)
  }

  get path(): SimplePath {
    const [x, y] = this.at
    return SimplePath.withPoints([
      this.at,
      [x + this.w, y],
      [x + this.w, y + this.h],
      [x, y + this.h],
    ]).close()
  }

  split = (config: {
    orientation: "vertical" | "horizontal"
    split?: number | number[]
  }): Rect[] => {
    let { orientation, split } = config
    split = split || 0.5
    if (orientation === "horizontal") {
      if (typeof split === "number") {
        return [
          new Rect({ at: this.at, w: this.w / 2, h: this.h }),
          new Rect({
            at: [this.at[0] + this.w / 2, this.at[1]],
            w: this.w / 2,
            h: this.h,
          }),
        ]
      } else {
        const total = split.reduce((a, b) => a + b, 0)
        const proportions = split.map((s) => s / total)
        let xDxs: [number, number][] = []
        let c = 0
        proportions.forEach((p) => {
          xDxs.push([c, p * this.w])
          c += p * this.w
        })
        return xDxs.map(
          ([x, dX], i) =>
            new Rect({
              at: [this.at[0] + x, this.at[1]],
              w: dX,
              h: this.h,
            })
        )
      }
    } else {
      if (typeof split === "number") {
        return [
          new Rect({ at: this.at, w: this.w, h: this.h / 2 }),
          new Rect({
            at: [this.at[0], this.at[1] + this.h / 2],
            w: this.w,
            h: this.h / 2,
          }),
        ]
      } else {
        const total = split.reduce((a, b) => a + b, 0)
        const proportions = split.map((s) => s / total)
        let yDys: [number, number][] = []
        let c = 0
        proportions.forEach((p) => {
          yDys.push([c, p * this.h])
          c += p * this.h
        })
        return yDys.map(
          ([y, dY], i) =>
            new Rect({
              at: [this.at[0], this.at[1] + y],
              w: this.w,
              h: dY,
            })
        )
      }
    }
  }
}
