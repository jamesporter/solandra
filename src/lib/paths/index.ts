export interface Traceable {
  traceIn(ctx: CanvasRenderingContext2D): void
}
