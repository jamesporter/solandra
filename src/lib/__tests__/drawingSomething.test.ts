import { describe, it, expect } from 'vitest';

import SCanvas from "../sCanvas"
import { Rect } from "../paths/Rect"

describe("Draw something with Solandra", () => {
  it("should be able to draw a rectangle", () => {
    const history: string[] = []
    const pocFakeCanvas = new Proxy(
      {},
      {
        get: function (target, property, receiver) {
          history.push(`${String(property)}`)
          return (...args: any[]) => {
            history[history.length - 1] = `${String(property)}(${args.join(
              ", "
            )})`
          }
        },
        set: function (target, property, value, receiver) {
          history.push(`${String(property)} = ${value}`)
          return true
        },
      }
    )

    const s = new SCanvas(
      pocFakeCanvas as CanvasRenderingContext2D,
      { width: 100, height: 100 },
      1,
      0
    )
    s.background(100, 20, 20)
    s.draw(new Rect({ at: [0.2, 0.2], w: 0.4, h: 0.2 }))
    expect(history).toMatchSnapshot()
  })
})
