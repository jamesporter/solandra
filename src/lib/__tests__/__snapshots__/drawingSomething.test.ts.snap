// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw something with Solandra should be able to draw a rectangle 1`] = `
Array [
  "resetTransform()",
  "scale(100, 100)",
  "lineWidth = 0.01",
  "lineJoin = round",
  "strokeStyle = black",
  "fillStyle = gray",
  "lineCap = round",
  "lineJoin = round",
  "save()",
  "fillStyle = hsla(100, 20%, 20%, 1)",
  "beginPath()",
  "rect(0, 0, 1, 1)",
  "fill()",
  "restore()",
  "beginPath()",
  "rect(0.2, 0.2, 0.4, 0.2)",
  "stroke()",
]
`;
